core/src/main/java/com/linecorp/armeria/internal/server/annotation/FileAggregatedMultipart.java

```diff
@@ -81,17 +98,43 @@
         });
     }
 
+    private static CompletableFuture<Path> moveFile(Path file, Path targetDirectory,
+                                                    ExecutorService blockingExecutorService) {
+        return CompletableFuture.supplyAsync(() -> {
+            createDirectories(targetDirectory);
+            try {
+                // Avoid name duplication, create new file at target place and replace it.
+                return Files.move(file, Files.createTempFile(targetDirectory, null, ".multipart"),
+                                  StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }, blockingExecutorService);
+    }
+
     private static CompletableFuture<Path> resolveTmpFile(Path directory,
                                                           ExecutorService blockingExecutorService) {
         return CompletableFuture.supplyAsync(() -> {
+            createDirectories(directory);
             try {
-                if (!Files.exists(directory)) {
-                    Files.createDirectories(directory);
-                }
                 return Files.createTempFile(directory, null, ".multipart");
             } catch (IOException e) {
                 throw new UncheckedIOException(e);
             }
         }, blockingExecutorService);
     }
+
+    private static void createDirectories(Path directory) {
+        directoryCreationCache.get(directory.toString(),
```

Comment:
Not sure if `Files.createDirectories` is thread safe or not. So wrap with caffeine to keep it trigger once.(I think creating directory when really using it instead of creating directory before starting Server)

Reply:
`createDirectories` uses `createDirectory` whose Javadoc notes that is atomic.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/Files.html#createDirectory(java.nio.file.Path,java.nio.file.attribute.FileAttribute...)
Consequently, `createDirectories` is atomic? How about removing the cache?
`Files.exists` guard looks good to me.